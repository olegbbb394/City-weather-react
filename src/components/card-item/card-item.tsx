import React, {FC, JSX} from 'react';
import {Link} from "react-router-dom";
import {Box, Button, Card, CardActions, CardContent, CardMedia, Typography} from "@mui/material";
import Grid from "@mui/material/Unstable_Grid2";

import {IWeather, ICityWeatherItem} from "../../types/city-weather.model";
import {LocalStorage} from "../../constants/localStorage-settings.enum";

import {getWeatherAction} from "../../redux/actions/get-weather.action";
import {useTypeDispatch} from "../../hooks/use-type-dispatch";
import {setWeatherDataArr} from "../../redux/reducers/weatherSlice";

type Props = {
    item: ICityWeatherItem;
}

export const CardItem: FC<Props> = ({item}: Props): JSX.Element => {

    if (!item) {
        return <></>
    }

    const dispatch = useTypeDispatch();

    const savingDataSelectedCard = (id: number) => {
        if (item?.id === id) {
            localStorage.setItem(LocalStorage.localStorageCardKey, JSON.stringify(item));
        }
    };

    const deleteCard = (id: number): void => {
        const dataArr = localStorage.getItem(LocalStorage.localStorageCardsKey);
        const filterArr = dataArr ? JSON.parse(dataArr)?.filter((item: ICityWeatherItem) => item.id !== id) : [];
        localStorage.setItem(LocalStorage.localStorageCardsKey, JSON.stringify(filterArr))
        dispatch(setWeatherDataArr(filterArr));
    };

    const updateCard = (name: string) => {
        if (name) dispatch(getWeatherAction(name));
    };

    return (
        <Grid xs={12} md={2}>
            <Card sx={{width: 250, border: '1px solid #4caf50', borderRadius: 2}}>
                <Link to='/card' onClick={() => savingDataSelectedCard(item.id)}>
                    <CardContent>
                        <Typography gutterBottom variant="h4" component="h4">
                            {item.name}
                        </Typography>
                        <Box sx={{width: 50, height: 50}}>
                            <CardMedia
                                component="img"
                                alt="weather icon"
                                image={`https://openweathermap.org/img/wn/${item.weather.map((item) => item.icon)}@2x.png`}
                            />
                        </Box>
                        <Typography variant="h5" component="h5">
                            {Math.round(item.main.temp)}&deg;
                        </Typography>
                        <Typography variant="body1" color="text.secondary">
                            pressure: {item.main.pressure} {'hPa'}
                        </Typography>
                        <Typography variant="body1" color="text.secondary">
                            description: {item.weather.map((item: IWeather) => item.description)}
                        </Typography>
                        <Typography variant="body1" color="text.secondary">
                            main: {item.weather.map((item: IWeather) => item.main)}
                        </Typography>
                        <Typography variant="body1" color="text.secondary">
                            wind: {item.wind.speed} {'km/h'}
                        </Typography>
                    </CardContent>
                </Link>
                <CardActions>
                    <Button size="small" variant="outlined" color='success' sx={{mr: 1}}
                            onClick={() => deleteCard(item.id)}>Delete
                        card</Button>
                    <Button size="small" variant="outlined" color='success' onClick={() => updateCard(item.name)}>Update
                        card</Button>
                </CardActions>
            </Card>
        </Grid>
    );
};
