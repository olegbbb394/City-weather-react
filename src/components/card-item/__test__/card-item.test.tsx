import {render, screen, act} from "../../../test-utils";
import {BrowserRouter as Router} from 'react-router-dom';
import userEvent from "@testing-library/user-event";
import {CardItem} from "../card-item";
import {ICityWeatherItem} from "../../../types/city-weather.model";
import {getWeatherAction} from "../../../redux/actions/get-weather.action";

describe('Card item component', () => {

    it('should render Card item component with data', () => {

        const item: ICityWeatherItem | null = {
            id: 703448,
            main: {pressure: 1000, temp: 35},
            name: 'Kyiv',
            weather: [{description: 'good', icon: 'o2d', id: 2, main: 'good'}],
            wind: {speed: 20}
        };

        const cardItemData = render(
            <Router basename="/">
                <CardItem item={item}/>
            </Router>
        );

        expect(screen.getByRole('img')).toBeInTheDocument();
        expect(screen.getByRole('link')).toBeInTheDocument();
        expect(screen.getByText(item.name)).toBeInTheDocument();
        expect(cardItemData).toMatchSnapshot();
    });

    it('should button delete card,deletes the card', async () => {

        let item: ICityWeatherItem | null = {
            id: 703448,
            main: {pressure: 1000, temp: 35},
            name: 'Kyiv',
            weather: [{description: 'good', icon: 'o2d', id: 2, main: 'good'}],
            wind: {speed: 20}
        };
        const getDeleteButton = () => {
            const button = screen.getByRole('button', {name: /Delete card/i});
            item = null;
            return button;
        };

        render(
            <Router basename="/">
                <CardItem item={item}/>
            </Router>
        );

        expect(item.name).toBeTruthy();
        expect(getDeleteButton()).toBeInTheDocument();

        await act(async () => {
            await userEvent.click(getDeleteButton());
        });

        expect(item).toBeNull();
    });

    it('should button update card,updates the card', async () => {
        const dispatchMock = jest.fn();
        const getUpdateButton = () => screen.getByRole('button', {name: /Update card/i});

        const item: ICityWeatherItem | null = {
            id: 703448,
            main: {pressure: 1000, temp: 35},
            name: 'Kyiv',
            weather: [{description: 'good', icon: 'o2d', id: 2, main: 'good'}],
            wind: {speed: 20}
        };

        const getWeatherActionMock = getWeatherAction(item.name);

        render(
            <Router basename="/">
                <CardItem item={item}/>
            </Router>
        );

        await act(async () => {
            await userEvent.click(getUpdateButton());
        });

        dispatchMock(getWeatherActionMock);

        expect(getUpdateButton()).toBeInTheDocument();
        expect(dispatchMock).toHaveBeenCalledTimes(1);
    });

    it('should saving  Selected Card data in localStorage', async () => {

        const localStorageSpy = jest.spyOn(window.localStorage.__proto__, 'setItem');

        const item: ICityWeatherItem | null = {
            id: 703448,
            main: {pressure: 1000, temp: 35},
            name: 'Kyiv',
            weather: [{description: 'good', icon: 'o2d', id: 2, main: 'good'}],
            wind: {speed: 20}
        };

        render(
            <Router basename="/">
                <CardItem item={item}/>
            </Router>
        );

        const link = screen.getByRole('link');

        expect(link).toBeInTheDocument();
        expect(localStorageSpy).not.toHaveBeenCalled();
        expect(item.name).toBeTruthy();

        await act(async () => {
            await userEvent.click(link);
        });

        expect(localStorageSpy).toHaveBeenCalledTimes(1);
        expect(localStorageSpy).toMatchSnapshot();
    });
});