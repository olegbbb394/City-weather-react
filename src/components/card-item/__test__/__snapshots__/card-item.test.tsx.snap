// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card item component should render Card item component with data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-2 css-tcbi16-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-i1c8im-MuiPaper-root-MuiCard-root"
        >
          <a
            href="/card"
          >
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <h4
                class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1kqmfgr-MuiTypography-root"
              >
                Kyiv
              </h4>
              <div
                class="MuiBox-root css-1u7uhqd"
              >
                <img
                  alt="weather icon"
                  class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                  src="https://openweathermap.org/img/wn/o2d@2x.png"
                />
              </div>
              <h5
                class="MuiTypography-root MuiTypography-h5 css-gvjb6m-MuiTypography-root"
              >
                35
                °
              </h5>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                pressure: 
                1000
                 
                hPa
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                description: 
                good
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                main: 
                good
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                wind: 
                20
                 
                km/h
              </p>
            </div>
          </a>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1d53n18-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Delete card
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-7yf5rt-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Update card
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-2 css-tcbi16-MuiGrid2-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-i1c8im-MuiPaper-root-MuiCard-root"
      >
        <a
          href="/card"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1kqmfgr-MuiTypography-root"
            >
              Kyiv
            </h4>
            <div
              class="MuiBox-root css-1u7uhqd"
            >
              <img
                alt="weather icon"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                src="https://openweathermap.org/img/wn/o2d@2x.png"
              />
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-gvjb6m-MuiTypography-root"
            >
              35
              °
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
            >
              pressure: 
              1000
               
              hPa
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
            >
              description: 
              good
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
            >
              main: 
              good
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
            >
              wind: 
              20
               
              km/h
            </p>
          </div>
        </a>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1d53n18-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Delete card
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-7yf5rt-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Update card
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Card item component should saving  Selected Card data in localStorage 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "card",
      "{\\"id\\":703448,\\"main\\":{\\"pressure\\":1000,\\"temp\\":35},\\"name\\":\\"Kyiv\\",\\"weather\\":[{\\"description\\":\\"good\\",\\"icon\\":\\"o2d\\",\\"id\\":2,\\"main\\":\\"good\\"}],\\"wind\\":{\\"speed\\":20}}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
